### ISSUE TITLE
Adding the concepts graph

### ISSUE THREAD

**Issue by iHiD** [exercism/website#105]
The "Concepts" tab on an internal track page is going to be made up of a search/filtering bar and then a graph/dependency view of the Concepts, with different states. Along the lines of this:

<img width="1331" alt="Screenshot 2020-09-24 at 21 45 55" src="https://user-images.githubusercontent.com/286476/94198555-578fdf80-feaf-11ea-94a1-fad5e36cf87b.png">

- The graph will be rendered once on page load - then static (although it probably needs lines redrawing on window resize etc)
- The filters will change **opacity** of the boxes.
- Hovering on boxes will open a Popper modal, and change the opacity of connecting lines.

There are fundamentally four parts to this:
1. Implementing searching/filtering (probably best server side but changing the state of elements within the page, rather than replacing the elements)
2. Generating the graph layout
3. Drawing the graph and lines
4. Styling.

- @neenjaw has done most of the work for (2) already. @ErikSchierboom, @kntsoriano Could you liase with him to work out how best to PR that into the website - and whether it needs converting to JS (it's currently TS).
- @neenjaw: If you want to do any of the work on doing (3) properly, then I'm sure Erik and Karlo would be happy for the support. You may want to put up an early (saner) POC using React as a basis for discussion. If you don't want to take this on then Erik/Karlo can do it and get you to review it in reverse. Let us know 🙂 
- @kntsoriano, @ErikSchierboom I think (1) is a relatively straight-forward server integration based on the work so far. Tell me the ideal JSON structure and I'll work exactly what data is going in. Just work on the basis of Concept Name for now.

---

_Also note the [awful POC](https://github.com/exercism/v3-website/pull/104) I did for it 🙂_

**Comment by iHiD** [comment]
(@SleeplessByte When you're back from beating COVID I'm sure everyone would appreciate your wisdom here too)

**Comment by neenjaw** [comment]
I can make a leaner POC so that it is easier to see just what is happening in the code. Will probably need a day or two with other things going on.

**Comment by iHiD** [comment]
@neenjaw Awesome. Thanks!

**Comment by ErikSchierboom** [comment]
@neenjaw That would be great!

**Comment by ErikSchierboom** [comment]
> @kntsoriano, @ErikSchierboom I think (1) is a relatively straight-forward server integration based on the work so far. Tell me the ideal JSON structure and I'll work exactly what data is going in. Just work on the basis of Concept Name for now.

@neenjaw You have already worked with the data. What is the best format to use?

**Comment by neenjaw** [comment]
@ErikSchierboom, I think there are definitely two concerns here when it comes to the data format, which I think is somewhat determined by the front-end design of the track progression as the user sees it on the website:

1) Data to display the track procession/links
2) Data to support filtering/searching etc

As far as point 1 is concerned, data similar to the track config (pulled from the DB directly or otherwise) is probably good since [there is a POC to massage it into a display format](https://github.com/neenjaw/ruby-exercism-graph).

Point 2 is less known to me as it depends where exactly the filtering/searching will be done (server side or client side) and how that will affect the display of the track progression

**Comment by neenjaw** [comment]
Also, [here is my progress on a react POC for the track progression so far](https://github.com/neenjaw/react-v3-website-poc)

**Comment by neenjaw** [comment]
Ok, been thinking and playing around with this. My initial thoughts around the data was around having the current config.json's data passed to react, but ordered to represent the layers of the graph to structure it, like so:

```
[
  [
    {
      index: 0,
      slug: 'basics',
      uuid: 'c29c6092-9d44-4f21-8138-b873384fd90b',
      concepts: ['basics'],
      prerequisites: [],
      status: "unlocked",
    },
  ],
  [
    // ...
  ]
]
```

But I started hitting some edge cases where I was basically facing the choice to recreate what is already written on the ruby side. I'm concerned that continuing down that path will likely be unsuccessful to implement the design stuff elegantly. 

So I was thinking that if the react side can be supplied with the exercises (nodes) and the connections (edges) as they are precomputed and cached in ruby, it might make things a lot better.  It might appear like so:

```
{
  exercises: {
    0: {
      slug: "basics",
      status: unlocked,
    },
    // and so on
  },
  layout: [
    [0],
    [1, 2],
    // and so on
  ],
  edges: [
    {
      from: 0,
      to: 1
    },
    // and so on
  ]
}
```

Pros:
  - This data is already computed and cached by the backend
  - This makes it much easier to implement a front end since no processing of the graph needs to be done
    - It makes the connections easier to draw since they will all be known rather than having to be discovered

Cons:
  - Some transformation may be required on the backend to massage the data further than what it is presently

Further questions:
  - if exercises are to be further filtered, where/how is that metadata hung on to this payload

**Comment by neenjaw** [comment]
This might be even easier to pull in from the database, then the layout and edges can be cached, and the student's progress can simply be tacked on to it:
```
{
  exercises: [
    {
      slug: "basics",
      status: unlocked,
    },
    // and so on
  ],
  layout: [
    ["basics"],
    ["numbers", "strings"],
    // and so on
  ],
  edges: [
    {
      from: "basics",
      to: "numbers"
    },
    // and so on
  ]
}
```

**Comment by iHiD** [comment]
@neenjaw Yeah - that's how I'd approach it. I'd have one thing that just generated the graph. We'll want this anyway to show to people that aren't logged in. Then student's "states" can either be merged into that if appropriate, or they could be a seperate JSON object, with JS uses to do stuff.

Re the filtering, we'll probably hit the server when someone filters and get back a list of `[match: [basics, strings], no_match: numbers]` or something, which can then be used to do any recalcing? Maybe?

**Comment by neenjaw** [comment]
> @neenjaw Yeah - that's how I'd approach it. I'd have one thing that just generated the graph. We'll want this anyway to show to people that aren't logged in. Then student's "states" can either be merged into that if appropriate, or they could be a seperate JSON object, with JS uses to do stuff.

Okay, so I'll move forward with some variation of the second version, whether the student's progress is merged to the `exercises` key, or another key later shouldn't be difficult to change to. 

> Re the filtering, we'll probably hit the server when someone filters and get back a list of `[match: [basics, strings], no_match: numbers]` or something, which can then be used to do any recalcing? Maybe?

When in filter/search mode, do things still appear as a directed graph? Or does the display format change?

**Comment by iHiD** [comment]
> When in filter/search mode, do things still appear as a directed graph? Or does the display format change?

Yeah - it'll be exactly the same graph. I don't even expect we'd want to re-render. The only changes will be that the non-selected nodes will have an opacity change, and hovering will be disabled, which we can probably do by appending CSS classes.

**Comment by kntsoriano** [comment]
Can we close now?
